Funcky/Sequence.cs

```diff
@@ -1,14 +1,47 @@
+using System;
 using System.Collections.Generic;
 using System.Diagnostics.Contracts;
 using System.Linq;
 using Funcky.GenericConstraints;
+using Funcky.Monads;
 
 namespace Funcky
 {
     public static class Sequence
     {
         public static IEnumerable<TItem> Return<TItem>(TItem item) => Enumerable.Repeat(item, 1);
 
+        /// <summary>
+        /// Returns an endless <see cref="IEnumerable{T}"/> that yields values based on the previous value
+        /// using the <paramref name="next" /> function.
+        /// This is essentially the inverse operation of an <see cref="Enumerable.Aggregate{T}"/>.
+        /// </summary>
+        public static IEnumerable<TItem> Generate<TItem>(TItem seed, Func<TItem, TItem> next)
+            where TItem : notnull
+        {
+            var previousItem = seed;
+            while (true)
+            {
+                yield return previousItem = next(previousItem);
+            }
+        }
+
+        /// <summary>
+        /// Returns an <see cref="IEnumerable{T}"/> that yields values based on the previous value
+        /// using the <paramref name="next" /> function. It stops when <paramref name="next" /> returns a
+        /// <see cref="Option{TItem}.None"/> value.
+        /// This is essentially the inverse operation of an <see cref="Enumerable.Aggregate{T}"/>.
+        /// </summary>
+        public static IEnumerable<TItem> Generate<TItem>(TItem seed, Func<TItem, Option<TItem>> next)
+            where TItem : notnull
+        {
+            var previousItem = seed;
+            while (next(previousItem).TryGetValue(out var item))
```

Comment:
Hmm yes that would work. But it requires that `item` is unwrapped:
```csharp
var previousItem = seed;
for (var item = next(previousItem); item.Match(true, False); item = next(previousItem))
{
    yield return previousItem = item.GetOrElse(() => throw new InvalidOperationException());
}
```

Reply:
The functional approach would be injecting the behaviour, which is impossible because you cannot yield in a lambda.

On way would be to do it like this, which is a bit awkward, but would work.

```cs
var previousItem = seed;
for (var item = next(previousItem); item.Match(none: false, some: True); item = next(previousItem))
{
    foreach(var enumeratedItem in item.ToEnumerable()) 
    {
        yield return previousItem = enumeratedItem ;
    }
}
```

But there is another functional approach, function composition. We have here an endless Generator in which we are interested until it returns a none.

```cs
 public static IEnumerable<Option<TItem>> GenerateInternal<TItem>(TItem seed, Func<TItem, Option<TItem>> next)
```

 public static IEnumerable<Option<TItem>> Generate<TItem>(TItem seed, Func<TItem, Option<TItem>> next)
    => GenerateInternal(seed, next).TakeWhile(e => e.Match(false, True)).WhereSelect();

Reply:
All these solutions have excessive checks that work around not having `TryGetValue`. I can't really see any benefits from working around it.